### PR TITLE
 Improved convertReflectValueToType

### PR DESCRIPTION
**Pull request by MichaelS11**
Updated Interrupt tests
Added Interface check to DeleteExpr

### CODE REVIEW

File: vm/vmExprFunction.go [hunk 1]
```diff
@@ -148,7 +148,7 @@ func callExpr(callExpr *ast.CallExpr, env *Env) (rv reflect.Value, err error) {
 		for i, expr := range callExpr.SubExprs {
 			if addrExpr, ok := expr.(*ast.AddrExpr); ok {
 				if identExpr, ok := addrExpr.Expr.(*ast.IdentExpr); ok {
-					invokeLetExpr(identExpr, args[i].Elem().Elem(), env)
+					invokeLetExpr(identExpr, args[i].Elem(), env)
```
Comment (MichaelS11) [inline vm/vmExprFunction.go]: @mattn 
So this reverts the previous commit you did. May I ask what you were trying to fix? Can I add a test for what you were trying to fix?

I added a test that will cause a panic if this is not change.

File: vm/vm_test.go [hunk 2]
```diff
@@ -456,7 +465,7 @@ for {
 `,
 		`
 a = []
-for i = 0; i < 10000; i++ {
+for i = 0; i < 100000; i++ {
```
Comment (mattn) [inline vm/vm_test.go]: please separate changes for TestInterrupts

File: vm/vmExprFunction.go (hunk 1 above)
Comment (mattn) [reply vm/vmExprFunction.go]: first Elem() is for dereference the pointer-variable. second one is for dereference the value that was set via v.Set (should be reflect.Interface).

File: vm/vm_test.go (hunk 2 above)
Comment (MichaelS11) [reply vm/vm_test.go]: Reverted TestInterrupts count change

File: vm/vm_test.go [hunk 3]
```diff
@@ -536,10 +545,8 @@ func runInterruptTest(t *testing.T, script string) {
 	}()
 
 	_, err = env.Execute(script)
-	if err == nil {
-		t.Errorf("Execute error - received %v - expected: %v", err, InterruptError)
-	} else if err.Error() != InterruptError.Error() {
-		t.Errorf("Execute error - received %v - expected: %v", err, InterruptError)
+	if err == nil || err.Error() != InterruptError.Error() {
```
Comment (mattn) [inline vm/vm_test.go]: please fix conflict.

File: vm/vmExprFunction.go (hunk 1 above)
Comment (MichaelS11) [reply vm/vmExprFunction.go]: But it might not be a interface, right?

**Comment by MichaelS11** [comment]
Merge fixed

File: vm/vmExprFunction.go (hunk 1 above)
Comment (mattn) [reply vm/vmExprFunction.go]: sure.

File: vm/vm_test.go [hunk 4]
```diff
@@ -581,9 +584,7 @@ func TestInterruptConcurrency(t *testing.T) {
 	for i := 0; i < 100; i++ {
 		go func() {
 			_, err := env.Execute("for { }")
-			if err == nil {
-				t.Errorf("Execute error - received %v - expected: %v", err, ErrInterrupt)
-			} else if err.Error() != ErrInterrupt.Error() {
+			if err != ErrInterrupt {
```
Comment (mattn) [inline vm/vm_test.go]: ditto

File: vm/vm_test.go [hunk 5]
```diff
@@ -564,9 +569,7 @@ func TestInterruptConcurrency(t *testing.T) {
 	waitGroup.Wait()
 
 	_, err := env.Execute("for { }")
-	if err == nil {
-		t.Errorf("Execute error - received %v - expected: %v", err, ErrInterrupt)
-	} else if err.Error() != ErrInterrupt.Error() {
+	if err != ErrInterrupt {
```
Comment (mattn) [inline vm/vm_test.go]: ditto

File: vm/vm_test.go [hunk 6]
```diff
@@ -551,9 +558,7 @@ func TestInterruptConcurrency(t *testing.T) {
 	for i := 0; i < 100; i++ {
 		go func() {
 			_, err := env.Execute("for { }")
-			if err == nil {
-				t.Errorf("Execute error - received %v - expected: %v", err, ErrInterrupt)
-			} else if err.Error() != ErrInterrupt.Error() {
+			if err != ErrInterrupt {
```
Comment (mattn) [inline vm/vm_test.go]: ditto

File: vm/vm_test.go [hunk 7]
```diff
@@ -536,10 +545,8 @@ func runInterruptTest(t *testing.T, script string) {
 	}()
 
 	_, err = env.Execute(script)
-	if err == nil {
-		t.Errorf("Execute error - received %v - expected: %v", err, ErrInterrupt)
-	} else if err.Error() != ErrInterrupt.Error() {
-		t.Errorf("Execute error - received %v - expected: %v", err, ErrInterrupt)
+	if err == nil || err.Error() != ErrInterrupt.Error() {
```
Comment (mattn) [inline vm/vm_test.go]: Let's remove this part at this PR.

Comment (MichaelS11) [reply vm/vm_test.go]: May I ask why? All that is doing it removing an extra if statement.

Comment (mattn) [reply vm/vm_test.go]: It makes it difficult for me to find issue that problem related to a bug.

**Comment by mattn** [comment]
merged https://github.com/mattn/anko/pull/142

File: vm/vm_test.go (hunk 7 above)
Comment (MichaelS11) [reply vm/vm_test.go]: @mattn 

Would be happy to understand why this change makes it difficult to find the issue. If you could please let me know why, that would be great.

Here is what I could see:

The job https://travis-ci.org/mattn/anko/jobs/356702744 said:
--- FAIL: TestInterrupts (0.11s)
	vm_test.go:540: Execute error - received <nil> - expected: Execution interrupted

Can look at line 540 here:
https://github.com/mattn/anko/blob/72fa4855a49189e82452215a4cc4cb22b16b2b8c/vm/vm_test.go#L540

That line is in the function runInterruptTest, which will runs these tests:
https://github.com/mattn/anko/blob/72fa4855a49189e82452215a4cc4cb22b16b2b8c/vm/vm_test.go#L445-L507

So the change in the if statements will not make it any easier to figure out which test was the problem, right?

**Comment by MichaelS11** [comment]
@mattn 

Could you please let me know why some of this PR was dropped? Or if there is a general change on how you would like PRs to be done, opened up a issue about it.

**Comment by mattn** [comment]
Please do not feel wrong. I just want to separate PRs since I prefer that one PR is fix one issue. This is useful to find out what was wrong. I developed anko as I feel as. But now anko seems to be depended on some products. So I decided this. Please don't feel wrong meaning.

**Comment by MichaelS11** [comment]
No problem, happy to make PRs smaller when possible. Maybe next time it would be possible to share about the change ahead of time to make it easier to adapt?

**Comment by Chyroc** [comment]
Everyone is great

**Pull request closed without merging**